R/pmcmc.R

```diff
@@ -0,0 +1,428 @@
+  }
+  
+  inputs <- list(
+    data = data,
+    n_mcmc = n_mcmc,
+    pars = list(pars_obs = pars_obs, 
+                pars_init = pars_init, 
+                pars_min = pars_min, 
+                pars_max = pars_max, 
+                pars_sd = pars_sd,
+                pars_discrete = pars_discrete), 
+    n_particles = n_particles, 
+    steps_per_day = steps_per_day)
+  
+  
+  par_names <- c('beta', 'start_date')
+  
+  
+  
+  correct_format <- function(pars) {
+    is.list(pars) & setequal(names(pars), par_names)
+  }
+
+  if(!correct_format(pars_init)) {
+    stop("pars_init must be a list of length two with names 'beta' and 'start_date'")
+  }
+  if(!correct_format(pars_min)) {
+    stop("pars_min must be a list of length two with names 'beta' and 'start_date'")
+  }
+  if(!correct_format(pars_max)) {
+    stop("pars_max must be a list of length two with names 'beta' and 'start_date'")
+  }
+  if(!correct_format(pars_sd)) {
+    stop("pars_sd must be a list of length two with names 'beta' and 'start_date'")
+  }
+  if(!correct_format(pars_discrete)) {
+    stop("pars_discrete must be a list of length two with names 'beta' and 'start_date'")
+  }
+  
+  
+  is.numeric.list <- function(x) all(vapply(X = x, is.numeric, logical(1)))
+  is.logical.list <- function(x) all(vapply(X = x, is.logical, logical(1)))
+  
+  if(!is.numeric.list(pars_min)) {
+    stop('pars_min entries must be numeric')
+  }
+  if(!is.numeric.list(pars_max)) {
+    stop('pars_max entries must be numeric')
+  }
+  if(!is.numeric.list(pars_sd)) {
+    stop('pars_sd entries must be numeric')
+  }
+  if(!is.logical.list(pars_discrete)) {
+    stop('pars_discrete entries must be logical')
+  }
+  
+
+  n_pars <- length(par_names)
+  
+  # needs to be a vector to pass to reflecting boundary function
+  pars_min <- unlist(pars_min)  
+  pars_max <- unlist(pars_max)
+  pars_sd <- unlist(pars_sd)
+  pars_discrete <- unlist(pars_discrete)
+  
+  
+  
+  # create prior and likelihood functions given the inputs
+ 
+   if(is.null(log_prior)) {
+     # set improper, uninformative prior
+    log_prior <- function(pars) log(1e-10)
+   }
+  calc_lprior <- log_prior
+  
+  if(is.null(log_likelihood)) {
+    log_likelihood <- calc_loglikelihood
+  } else if (!('...' %in% names(formals(log_likelihood)))){
+    stop('log_likelihood function must be able to take unnamed arguments')
+  }
+  
+  
+  
+  # create shorthand function to calc ll given main inputs
+  calc_ll <- function(pars) {  
+    X <- log_likelihood(pars = pars, 
+                            data = data, 
+                            model_params = model_params,
+                            steps_per_day = steps_per_day, 
+                            pars_obs = pars_obs, 
+                            n_particles = n_particles
+    ) 
+    X
+  }
+  
+  
+  
+  # convert the current parameters into format easier for mcmc to deal with
+  curr_pars <- unlist(pars_init)
+  day_zero <- as.Date(data$date[1])
```

Comment:
How does this relate to the first data date, and the model start date? I guess it's day zero in the same sense as the data, but the model will run from prior to this. I think this is fine as it's consistent between this and the particle sampler, but might be useful to use a consistent variable name (see the different start date types in `run_particle_filter()`)

Reply:
I have made a couple of simplifying changes that should improve readability, including converting the data$date vector into Date class at the beginning of the code, rather than multiple times throughout, and removing the need for creating a 'day_zero' object